### ISSUE TITLE
[css-display][css-pseudo] Interaction of display:contents and ::first-line

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#1310]
Typically, inheritance follows the element tree. However, the [`::first-line`](https://drafts.csswg.org/css-pseudo-4/#first-line-inheritance) is able to "hijack" the inheritance of some properties:

> During CSS inheritance, the portion of a child element that occurs on the first line only inherits properties applicable to the ::first-line pseudo-element from the ::first-line pseudo-element.

The problem is that the place in which `::first-line` is generated depends on the box tree, and `display: contents` generates no boxes, so I'm not sure how they interact.

The implementations of `display: contents` (Firefox and Chrome) are not interoperable, and both seem undesirable:

 - On Firefox, inline contents in the first line are not affected by `::first-line` if they are inside an element with `display: contents`. They inherit directly from that element instead. [Example](https://jsfiddle.net/78wz9gee/). [Bug 1305951](https://bugzilla.mozilla.org/show_bug.cgi?id=1305951)
 - On Chrome, the mere existence of `::first-line` prevents inline contents in the first line from inheriting from a `display: contents` parent. They inherit first from `::first-line`, and then from the nearest non-`display: contents` ancestor. [Example](https://jsfiddle.net/91r5336d/). [Bug 706316](https://bugs.chromium.org/p/chromium/issues/detail?id=706316)

For example, https://jsfiddle.net/0q8paowL/

```css
main::first-line { color: blue; text-decoration: underline; }
div { display: contents; color: green; letter-spacing: 10px; }
span { display: contents; color: red; }
```
```html
<main>
  <div>aaa <span>bbb</span> <p>ccc</p></div>
</main>
```

I will attempt to represent the inheritance using a [fictional tag sequence](https://drafts.csswg.org/css-pseudo-4/#fictional-tag-sequence), which I'm not sure is right because `::first-line` only exists in the box tree and `display: contents` in the element tree.

I see three reasonable possibilities:

 - Place the `::first-line` as outermost as possible, breaking `display: contents` elements if necessary:

    ```html
    <main>
      <main::first-line>
        <div no-box>aaa <span no-box>bbb</span></div>
      </main::first-line>
      <div><p>ccc</p></div>
    </main>
    ```

    Then `aaa` would be green (from `div`) and `bbb` would be red (from `span`). Both would inherit `letter-spacing` (from `div`). This would be closer to what Firefox does, but they would also receive a blue underline from `::first-line`.

 - Nest `::first-line` inside `display: contents` elements, breaking it if necessary:

    ```html
    <main>
      <div no-box>
        <main::first-line>aaa</main::first-line>
        <span no-box><main::first-line>bbb</main::first-line></span>
        <p>ccc</p>
      </div>
    </main>
    ```

    Then `aaa` would be blue (from `::first-line`) and `bbb` would be blue (from `::first-line`). Both would have blue underline (from `::first-line`). This would be closer to what Chrome does, but they would also inherit `letter-spacing` (from `div`).

 - Let `::first-line` enclose `display: contents` elements that only have inline contents, but nest it inside `display: contents` elements that only have some block-level.

    ```html
    <main>
      <div no-box>
        <main::first-line>aaa <span no-box>bbb</span></main::first-line>
        <p>ccc</p>
      </div>
    </main>
    ```

    Then `aaa` would be blue (from `::first-line`) and `bbb` would be red (from `span`). Both would have blue underline (from `::first-line`) and `letter-spacing` (from `div`). This would be like a mix of the previous possibilities.

**Comment by Loirooriol** [comment]
I think the first option makes more sense, because the others may need to break `::first-line`, which until now I think it was never necessary. And `display: contents` seems more inliney than blocky to me, so nesting it inside `::first-line` makes sense. This is also the behavior which would result from the alternative definition of `display: contents` that I proposed in #1313.

**Comment by tabatkins** [comment]
Firefox is, as far as I can tell, broken in general for ::first-line on an ancestor. In particular, the following:

```
<!DOCTYPE html>
<style>
main::first-line { color: blue; text-decoration: underline; }
div { color: green; letter-spacing: 10px; }
span { color: red; }
</style>
<main>
  <div>aaa <span>bbb</span> <p>ccc</p></div>
</main>
```
displays  "aaa" in green, and nothing with any underlines, at least in v53.

**Comment by Loirooriol** [comment]
Yes, that's [bug 317081](https://bugzilla.mozilla.org/show_bug.cgi?id=317081).

**Comment by tabatkins** [comment]
Yes, @fantasai and I think that the first option is best.  #2 is bad because it produces sibling ::first-line boxes, which doesn't happen currently and would break certain types of formatting, such as image backgrounds and letter-spacing. 

#3 is bad because it's actually impossible - the boxes don't exist, so we don't know whether they "would contain" inlines or blocks (that fact can change depending on surroundings).  Also, appending content to an element shouldn't fundamentally change the structure of earlier contents.

So #1 is the most reasonable, I think - just look at the resulting box tree and insert ::first-line boxes as appropriate.  This keeps ::first-line generation purely a box-tree-time operation (which both #2 and #3 fail, since they require element-tree information)  The only downside is that, from an author's perspective, it "flips" inheritance for text inside a display:contents that would have been a block (without display:contents, the ::first-line goes inside the div, so text inherits from the ::first-line rather than the div), however we don't really have a concept of "would have been a block" (as the block-ness comes from the display value).

**Comment by tabatkins** [comment]
AKA this bug depends on #1118 getting fixed first.

**Comment by Loirooriol** [comment]
Now I'm thinking that option 1 may be a bit more costly to implement, e.g. if you have

```html
<section>
  <div ><div ><div ><div ><div ><div ><div ><div ><div ><div >
    <span>Foo</span>
  </div></div></div></div></div></div></div></div></div></div>
</section>
```
```css
section::first-line { color: green }
div { display: contents }
```

When the browser generates the styles for the `section`, it sees a `::first-line` must be generated, but does not know where. I guess it must first iterate all the nested `div` because there could be some block container inside them. When it finds out that nope, then it goes back and creates a style for the outermost `div` which inherits from `section::first-line`, then a style for the second `div` which inherits from the previous style, and so on. Alternatively it could start creating these styles once it sees the first `display: contents` element, but if there were a block container inside, that work would have been useless.

Option 2 seems easier to implement: once you find a first-child inline-level inside a block container, create a `::first-line` fragment, let it inherit from the parent element of the inline, and place the inline-level (and following inline-level siblings, until the end of the line) inside that `::first-line`.

**Comment by fantasai** [comment]
CC @bzbarsky

**Comment by bzbarsky** [comment]
I don't see why the scenario described in https://github.com/w3c/csswg-drafts/issues/1310#issuecomment-303221414 is really a problem any more so than it would be if all the `<div>`s were `display:inline`.

Past that:

1. I wish the CSS WG would stop introducing features if they have not actually through through their interactions with all the features they have introduced.  The burden should be to prove that a new feature is compatible with all existing ones before it's added, not to discover during implementation that things are incompatible with each other.

2. I haven't really thought about how all this should interact with the multiple first-lines thing that CSS specifies (and no one actually implements in practice, at least not as specified).  As noted above, Gecko doesn't implement that, has no plans to implement it, and hence I don't have any good implementation experience or intuition about how it should work with `display:contents`.

3.  As noted, option 2 violates a current "there is at most one first-line box generated by a given selector for a given block" invariant which may lead to weirdness of all sorts.  The background situation is particularly troubling.

**Comment by Loirooriol** [comment]
@bzbarsky If the `<div>`s were `display:inline`, then their first fragments would inherit from `section::first-line` for sure. But with `display:contents` and option 1, you can't know if the pseudo-element will be placed outside or inside the `<div>`s until you check whether all child boxes are inline-level or there is some block-level.

That's how I see it, but I'm not an implementator, so I might be wrong or it might not be a big problem.

**Comment by bzbarsky** [comment]
> then their first fragments would inherit from section::first-line for sure.

Sure, but then you still have to recursively change the styles on them going down the tree...

That said, what Gecko actually implements is done on the box tree; so for us option 2 would in fact be quite a bit easier to implement, but not because of the "have to look at them all" bit.  Rather it's because it avoids the need to attach multiple styles to a single element (the display:contents thing), which is not something we have infrastructure for right now (we can attach styles to boxes, and one style to an element).

**Comment by bzbarsky** [comment]
That said, "quite a bit easier to implement" is still "pretty annoying to implement", since it would change other invariants...

**Comment by fantasai** [comment]
@Loirooriol I think the spec at this point would be implying your option 1, more or less. Do you see anything that needs further clarification, or should we close the issue?

**Comment by Loirooriol** [comment]
@fantasai I don't see where the spec implies option 1. It still seems unclear to me.
